### PR TITLE
Add tests for `BinaryReader` and `BinaryWriter`

### PR DESCRIPTION
**Pull request by Argmaster**
This pull request adds unit tests for `BinaryReader` and `BinaryWriter` which verify that data does not change after being serialized and deserialized. Tests cover significant edge cases present in possible inputs.

Resolves: #1262

### CODE REVIEW

File: src/utils.zig [hunk 1]
```diff
@@ -1530,6 +1530,228 @@ pub const BinaryWriter = struct {
 	}
 };
 
+const ReadWriteTest = struct {
+	fn init() void {
+		main.initThreadLocals();
+	}
+	fn deinit() void {
+		main.deinitThreadLocals();
+	}
```
Comment (IntegratedQuantum) [inline src/utils.zig:1539]: Instead of requiring these for every test (and every future test for that matter), how about just assigning the `stack`/`globalAllocator`s in main.zig the testing allocator if we are in a test build. You can use `if(builtin.is_test)` for this.

File: src/utils.zig [hunk 2]
```diff
@@ -1530,6 +1530,228 @@ pub const BinaryWriter = struct {
...
+test "read/write float" {
+	ReadWriteTest.init();
+	defer ReadWriteTest.deinit();
+
+	inline for([_]type{f16, f32, f64, f80, f128}) |floatT| {
+		try ReadWriteTest.testFloat(floatT, std.math.floatMax(floatT));
+		try ReadWriteTest.testFloat(floatT, 0.0012443);
+		try ReadWriteTest.testFloat(floatT, 0.0);
+		try ReadWriteTest.testFloat(floatT, 6457.0);
+		try ReadWriteTest.testFloat(floatT, std.math.floatMin(floatT));
```
Comment (IntegratedQuantum) [inline src/utils.zig:1653]: infinity would also be nice I think

File: src/utils.zig [hunk 3]
```diff
@@ -1530,6 +1530,228 @@ pub const BinaryWriter = struct {
...
+			std.math.floatMax(@typeInfo(vecT).vector.child),
+		});
+	}
+}
+
+test "read/write mixed" {
+	ReadWriteTest.init();
+	defer ReadWriteTest.deinit();
+
+	const type_first = u4;
```
Comment (IntegratedQuantum) [inline src/utils.zig:1723]: Should be camel case.
Also why not just use numbers? `type0`, `type1`...

File: src/utils.zig [hunk 4]
```diff
@@ -1530,6 +1530,228 @@ pub const BinaryWriter = struct {
...
+
+	try ReadWriteTest.testInt(u1, 0);
+	try ReadWriteTest.testInt(u1, 1);
+
+	inline for([_]type{u2, u4, u5, u8, u16, u31, u32, u64, u128}) |intT| {
+		const max = std.math.maxInt(intT);
+		std.debug.assert(max != 0);
+
+		const mid = (std.math.maxInt(intT) + std.math.minInt(intT))/2;
+		std.debug.assert(mid != 0);
```
Comment (IntegratedQuantum) [inline src/utils.zig:1607]: What's with the asserts here? Without this pointless assert you could test `u1` and `u0` here too.

File: src/utils.zig [hunk 5]
```diff
@@ -1530,6 +1530,228 @@ pub const BinaryWriter = struct {
...
+		std.debug.assert(min != 0);
+
+		const lowerMid = std.math.minInt(intT)/2;
+		std.debug.assert(lowerMid != 0);
+
+		const upperMid = std.math.maxInt(intT)/2;
+		std.debug.assert(upperMid != 0);
+
+		const max = std.math.maxInt(intT);
+		std.debug.assert(max != 0);
```
Comment (IntegratedQuantum) [inline src/utils.zig:1634]: What's with the asserts here? Without this pointless assert you could test `i2` (and i1) here in the loop too.

File: src/utils.zig [hunk 6]
```diff
@@ -1530,6 +1530,228 @@ pub const BinaryWriter = struct {
...
+		return .init(main.stackAllocator, .big);
+	}
+	fn getReader(data: []const u8) BinaryReader {
+		return .init(data, .big);
+	}
+	fn testInt(comptime intT: type, expected: intT) !void {
+		var writer = getWriter();
+		defer writer.deinit();
+		writer.writeInt(intT, expected);
+
```
Comment (IntegratedQuantum) [inline src/utils.zig:1550]: I would like you to also add an assert for the sizes, the size of the buffer after writing should always be `@ceilDiv(@bitSizeOf(intT), 8)`

File: src/utils.zig [hunk 7]
```diff
@@ -1530,6 +1530,228 @@ pub const BinaryWriter = struct {
...
+	writer.writeFloat(type_fourth, expected_fourth);
+	writer.writeSlice(expected_fifth);
+
+	var reader = ReadWriteTest.getReader(writer.data.items);
+
+	try std.testing.expectEqual(expected_first, try reader.readInt(type_first));
+	try std.testing.expectEqual(expected_second, try reader.readVec(type_second));
+	try std.testing.expectEqual(expected_third, try reader.readEnum(type_third));
+	try std.testing.expectEqual(expected_fourth, try reader.readFloat(type_fourth));
+	try std.testing.expectEqualStrings(expected_fifth, try reader.readSlice(expected_fifth.len));
```
Comment (IntegratedQuantum) [inline src/utils.zig:1752]: For the mixed test I would like to also see an is empty assert at the end. For the others I don't mind it, but here it is relevant in my opinion.

**Review by IntegratedQuantum — CHANGES_REQUESTED** [review]
It would be nice to see some error paths, e.g. an error for invalid enum values.

also feel free to reject some of my proposals here, I think I may have overdone it with the review here, it's just a test after all.

File: src/utils.zig (hunk 1 above)
Comment (Argmaster) [reply src/utils.zig:1539]: So I ended up doing:

```zig
const ReadWriteTest = struct {
	pub var testingAllocator = main.heap.ErrorHandlingAllocator.init(std.testing.allocator);
	pub var allocator = testingAllocator.allocator();
```

To completely avoid touching `main.zig`, is that alright?

File: src/utils.zig (hunk 6 above)
Comment (IntegratedQuantum) [reply src/utils.zig:1547]: Ah right, they never added that. There's std.math.divCeil though, and I'd appreciate you using it since it clearer conveys the intention here.

Comment (Argmaster) [reply src/utils.zig:1547]: Done

File: src/utils.zig (hunk 1 above)
Comment (IntegratedQuantum) [reply src/utils.zig]: For now it's fine, but this issue will come back if we actually test things that use these allocators. I made an issue for it: #1286

File: src/utils.zig [hunk 8]
```diff
@@ -1530,6 +1530,195 @@ pub const BinaryWriter = struct {
...
+
+		var reader = getReader(writer.data.items);
+		const actual = try reader.readVec(VecT);
+
+		try std.testing.expectEqual(expected, actual);
+	}
+};
+
+test "read/write unsigned int" {
+	inline for([_]type{u8, u1, u2, u4, u5, u8, u16, u31, u32, u64, u128}) |intT| {
```
Comment (IntegratedQuantum) [inline src/utils.zig:1596]: `u8` is listed twice, also would it be possible to add `u0`? It would be interesting to see how it behaves in the edge case of zero-sized types.

**Review by IntegratedQuantum — CHANGES_REQUESTED** [review]

File: src/utils.zig (hunk 8 above)
Comment (Argmaster) [reply src/utils.zig:1596]: That is definitely a typo

Comment (Argmaster) [reply src/utils.zig:1596]: Well, so behavior is basically that nothing is stored and nothing is red, which is kind of valid, but not very useful, is it?

Comment (IntegratedQuantum) [reply src/utils.zig:1596]: No it isn't, but it is an edge case that makes sense testing in my opinion.

Comment (IntegratedQuantum) [reply src/utils.zig]: `u0` can be e.g. the backing type of enums with only one entry.

**Review by IntegratedQuantum — APPROVED** [review]
